pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,389 @@
+	return b.client.DeleteObject(b.name, name)
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(_ context.Context, name string, r io.Reader) error {
+	size, err := objstore.TryToGetSize(r)
+	if err != nil {
+		return errors.Wrapf(err, "getting size of %s", name)
+	}
+
+	partNums, lastSlice := int(math.Floor(float64(size)/partSize)), size%partSize
+	if partNums == 0 {
+		body, err := bce.NewBodyFromSizedReader(r, lastSlice)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create SizedReader for %s", name)
+		}
+
+		if _, err := b.client.PutObject(b.name, name, body, nil); err != nil {
+			return errors.Wrapf(err, "failed to upload %s", name)
+		}
+
+		return nil
+	}
+
+	result, err := b.client.BasicInitiateMultipartUpload(b.name, name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to initiate MultipartUpload for %s", name)
+	}
+
+	uploadEveryPart := func(partSize int64, part int, uploadId string) (string, error) {
+		body, err := bce.NewBodyFromSizedReader(r, partSize)
+		if err != nil {
+			return "", err
+		}
+
+		etag, err := b.client.UploadPart(b.name, name, uploadId, part, body, nil)
+		if err != nil {
+			if err := b.client.AbortMultipartUpload(b.name, name, uploadId); err != nil {
+				return etag, err
+			}
+			return etag, err
+		}
+		return etag, nil
+	}
+
+	var parts []api.UploadInfoType
+
+	for part := 1; part <= partNums; part++ {
+		etag, err := uploadEveryPart(partSize, part, result.UploadId)
+		if err != nil {
+			return errors.Wrapf(err, "failed to upload part %d for %s", part, name)
+		}
+		parts = append(parts, api.UploadInfoType{PartNumber: part, ETag: etag})
+	}
+
+	if lastSlice != 0 {
+		etag, err := uploadEveryPart(lastSlice, partNums+1, result.UploadId)
+		if err != nil {
+			return errors.Wrapf(err, "failed to upload the last part for %s", name)
+		}
+		parts = append(parts, api.UploadInfoType{PartNumber: partNums + 1, ETag: etag})
+	}
+
+	if _, err := b.client.CompleteMultipartUploadFromStruct(b.name, name, result.UploadId, &api.CompleteMultipartUploadArgs{Parts: parts}); err != nil {
+		return errors.Wrapf(err, "failed to set %s upload completed", name)
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, opt ...objstore.IterOption) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, objstore.DirDelim) + objstore.DirDelim
+	}
+
+	delimiter := objstore.DirDelim
+
+	if objstore.ApplyIterOptions(opt...).Recursive {
+		delimiter = ""
+	}
+
+	var marker string
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		objects, err := b.client.ListObjects(b.name, &api.ListObjectsArgs{
+			Delimiter: delimiter,
+			Marker:    marker,
+			MaxKeys:   1000,
+			Prefix:    dir,
+		})
+
```

Comment:
```suggestion
```

nit: It's useful to bring err and err check together without a new line.

Reply:
> Amazing. LGTM, only suggestion towards documentation nits.
> 
> Thanks!

Ok thanks, I have updated the documentation.